gusto/state.py

```diff
@@ -123,36 +123,47 @@ def __init__(self, filename, diagnostics, description, create=True):
         if not create:
             return
-        with Dataset(filename, "w") as dataset:
-            dataset.description = "Diagnostics data for simulation {desc}".format(desc=description)
-            dataset.history = "Created {t}".format(t=time.ctime())
-            dataset.source = "Output from Gusto model"
-            dataset.createDimension("time", None)
-            var = dataset.createVariable("time", np.float64, ("time", ))
-            var.units = "seconds"
-            for name in diagnostics.fields:
-                group = dataset.createGroup(name)
-                for diagnostic in diagnostics.available_diagnostics:
-                    group.createVariable(diagnostic, np.float64, ("time", ))
+        if self.rank == 0:
+            with Dataset(filename, "w") as dataset:
+                dataset.description = "Diagnostics data for simulation {desc}".format(desc=description)
+                dataset.history = "Created {t}".format(t=time.ctime())
+                dataset.source = "Output from Gusto model"
+                dataset.createDimension("time", None)
+                var = dataset.createVariable("time", np.float64, ("time", ))
+                var.units = "seconds"
+                for name in diagnostics.fields:
+                    group = dataset.createGroup(name)
+                    for diagnostic in diagnostics.available_diagnostics:
+                        group.createVariable(diagnostic, np.float64, ("time", ))
 
     def dump(self, state, t):
         """Dump diagnostics.
 
         :arg state: The :class:`State` at which to compute the diagnostic.
         :arg t: The current time.
         """
-        with Dataset(self.filename, "a") as dataset:
-            idx = dataset.dimensions["time"].size
-            dataset.variables["time"][idx:idx + 1] = t
+# All MPI processes need to participate in calculating diagnostics but only the
+# rank 0 process writes to the netCDF file
+        if self.rank == 0:
+            with Dataset(self.filename, "a") as dataset:
+                idx = dataset.dimensions["time"].size
+                dataset.variables["time"][idx:idx + 1] = t
+                for name in self.diagnostics.fields:
+                    field = state.fields(name)
+                    group = dataset.groups[name]
+                    for dname in self.diagnostics.available_diagnostics:
+                        diagnostic = getattr(self.diagnostics, dname)
+                        var = group.variables[dname]
+                        var[idx:idx + 1] = diagnostic(field)
+        else:
             for name in self.diagnostics.fields:
                 field = state.fields(name)
-                group = dataset.groups[name]
                 for dname in self.diagnostics.available_diagnostics:
                     diagnostic = getattr(self.diagnostics, dname)
-                    var = group.variables[dname]
-                    var[idx:idx + 1] = diagnostic(field)
+                    diagnostic(field)
```

Comment:
For clarity, I would prefer having this calculation of diagnostics happen up front, and then dump on rank 0. So like this perhaps:

```python
diagnostics = []
for name in self.diagnostics.fields:
    field = state.fields(name)
    for dname in self.diagnostics.available_diagnostics:
        diagnostic = getattr(self.diagnostics, dname)
        diagnostics.append((name, dname, diagnostic(field))

if self.comm.rank == 0:
    with DataSet(...) as dataset:
        idx = dataset.dimensions["time"].size
        dataset.variables["time"][idx:idx+1] = t
        for field_name, name, diagnostic in diagnostics:
            group = dataset.groups[field_name]
            var = group.variables[name]
            var[idx:idx+1] = diagnostic
```

(untested)